client/src/Auth/TabPanel.tsx

```diff
@@ -0,0 +1,152 @@
+import React, { useState } from "react";
+import Typography from "@material-ui/core/Typography";
+import TextField from "@material-ui/core/TextField";
+import { Button } from "@material-ui/core";
+
+import { useStyles } from "./Auth.components";
+
+interface Props {
+  activeTab: number;
+  index: number;
+}
+
+const TabPanel: React.FC<Props> = ({ activeTab, index }) => {
+  const classes = useStyles();
+
+  const [credentials, setCredentials] = useState({
+    fullName: "",
+    username: "",
+    password: "",
+    confirmPassword: ""
+  });
+
+  const [error, setError] = useState("");
+
+  const fullName = credentials.fullName;
+  const username = credentials.username;
+  const password = credentials.password;
+  const confirmPassword = credentials.confirmPassword;
+
+  const isValidUsername = !!username && !/^[0-9a-zA-Z_.-]+$/.test(username);
+  const isValidPassword = !!password && password.length < 5;
+  const isValidConfirmedPassword =
+    !!confirmPassword && confirmPassword !== password;
+
+  const handleChange = (prop: string) => (
+    event: React.ChangeEvent<HTMLInputElement>
+  ) => {
+    setCredentials({ ...credentials, [prop]: event.target.value });
+  };
+
+  const register = async (event: React.FormEvent) => {
+    event.preventDefault();
+    if (!username || isValidUsername) {
+      setError("Username is invalid");
+    } else if (!password || isValidPassword) {
+      setError("Password is invalid");
+    } else if (!confirmPassword || isValidConfirmedPassword) {
+      setError(`Passwords don't match`);
+    } else {
+      setError("");
+    }
+  };
+
+  const login = async (event: React.FormEvent) => {
+    event.preventDefault();
+    if (!username || isValidUsername) {
+      setError("Username is invalid");
+    } else if (!password || isValidPassword) {
+      setError("Password is invalid");
+    } else {
+      setError("");
+    }
+  };
+
+  return (
+    <Typography
+      component="div"
+      role="tabpanel"
+      hidden={activeTab !== index}
+      id={`full-width-tabpanel-${index}`}
+      aria-labelledby={`full-width-tab-${index}`}
+    >
+      <form
+        className={classes.form}
+        noValidate
+        onSubmit={index === 1 ? register : login}
```

Comment:
Instead of this magic number 1 can we define a constant for it like
const LOGIN_TAB_INDEX = 1;